### PR TITLE
hack in allow_filtering

### PR DESCRIPTION
**Pull request by phoet**
for reference: https://github.com/cequel/cequel/issues/208#issuecomment-273941857

this is a working prototype for adding `ALLOW FILTERING` support. the way that i squeezed it into the existing codebase will make it work like `Post.allow_filtering!.where(title: 'Cequel 0')`. 

`allow_filtering!` will have to be called (like ie `consitency(:all)`) before calling `where`. this is mainly due to the `ArgumentError` that will otherwise be thrown in `filter_secondary_index`.

that does not feel very sexy to me though...

i'd like to get some early feedback before proceeding.

### CODE REVIEW

**Comment by phoet** [comment]
i was wondering if it might be possible to implement this via a global configuration setting ie `Cequel.allow_filtering!`.

with this, i could enable it for ie. a debugging session or in the development environment or if "i know what i'm doing".

i understand your concerns for trying to protect people from shooting themselves in the foot, but i don't see any difference to doing `Redis.keys` or a sql-query with no index...

File: lib/cequel/metal/data_set.rb [hunk 1]
```diff
@@ -573,6 +574,12 @@ def page_size(page_size)
         end
       end
 
+      def allow_filtering!
```
Comment (pezra) [inline lib/cequel/metal/data_set.rb]: I think a method comment would be helpful to explain what this is for and which C* version support it.

File: lib/cequel/record/data_set_builder.rb [hunk 2]
```diff
@@ -99,6 +101,12 @@ def set_consistency
         end
       end
 
+      def set_allow_filtering
```
Comment (pezra) [inline lib/cequel/record/data_set_builder.rb]: method docs here too

File: lib/cequel/record/record_set.rb [hunk 3]
```diff
@@ -465,6 +465,16 @@ def consistency(consistency)
       end
 
       #
+      # Set the consistency at which to read records into the record set.
+      #
+      # @param consistency [Symbol] consistency for reads
```
Comment (pezra) [inline lib/cequel/record/record_set.rb]: This line is incorrect.

File: lib/cequel/record/record_set.rb [hunk 4]
```diff
@@ -465,6 +465,16 @@ def consistency(consistency)
...
+      # @return [RecordSet] record set tuned to given consistency
+      #
```
Comment (pezra) [inline lib/cequel/record/record_set.rb]: Let's add a compatibility note here, too.

File: spec/examples/record/record_set_spec.rb [hunk 5]
```diff
@@ -805,6 +805,12 @@ def self.latest(count)
           to raise_error(ArgumentError)
       end
     end
+
+    context 'allow_filtering!' do
```
Comment (pezra) [inline spec/examples/record/record_set_spec.rb]: Tag this example so it will be skipped on old versions.

**Comment by phoet** [comment]
@pezra sorry, dont want to sound rude, did you actually read my comments?

**Comment by pezra** [comment]
@phoet, wondering if i read your comment is a fair question since i didn't respond directly.

Enabling `allow filtering` globally is much more dangerous than enabling it just for individual queries. In my experience `allow filtering` doesn't work very well for non-trivially sized tables. I am worried that those new to C* will enable allow filtering globally because it allows C* to behave more like a relational database. Doing so would probably be fine in development but will fail miserably if they take the app to production. 

I see the appeal for debugging and troubleshooting but i think the risk is too high for the limited benefit. Particularly, when the same effect is accomplished by this PR at much lower risk.

**Comment by pezra** [comment]
if you add the documentation to the methods and readme and get the build green i'd merge this today so that it is in the release i am planning this afternoon.

File: lib/cequel/record/data_set_builder.rb (hunk 2 above)
Comment (phoet) [inline lib/cequel/record/data_set_builder.rb]: all those methods are without doc, the builder is kind of a private api

**Comment by phoet** [comment]
@pezra i added some documentation. unfortunately i have to run now, so can't see if the build turns green or any other feedback. feel free to adapt anything that is not to your liking.

File: lib/cequel/record/data_set_builder.rb (hunk 2 above)
Comment (pezra) [inline lib/cequel/record/data_set_builder.rb]: i don't think internality is a particularly compelling excuse for omitting method docs. However, it is a small thing.

**Comment by pezra** [comment]
Merged. Thanks.

**Comment by pezra** [comment]
release in 2.1.0

**Comment by phoet** [comment]
<3

**Pull request merged**